bindings/python/unicorn/unicorn.py

```diff
@@ -1,492 +1,303 @@
-    _load_win_support('')
-
-def _load_lib(path, lib_name):
-    try:
-        if sys.platform in ('win32', 'cygwin'):
-            _load_win_support(path)
-
-        lib_file = os.path.join(path, lib_name)
-        dll = ctypes.cdll.LoadLibrary(lib_file)
-        #print('SUCCESS')
-        return dll
-    except OSError as e:
-        #print('FAIL to load %s' %lib_file, e)
-        return None
-
-_uc = None
-
-# Loading attempts, in order
-# - user-provided environment variable
-# - pkg_resources can get us the path to the local libraries
-# - we can get the path to the local libraries by parsing our filename
-# - global load
-# - python's lib directory
-# - last-gasp attempt at some hardcoded paths on darwin and linux
-
-_path_list = [os.getenv('LIBUNICORN_PATH', None),
-              pkg_resources.resource_filename(__name__, 'lib'),
-              os.path.join(os.path.split(__file__)[0], 'lib'),
-              '',
-              distutils.sysconfig.get_python_lib(),
-              "/usr/local/lib/" if sys.platform == 'darwin' else '/usr/lib64',
-              os.getenv('PATH', '')]
-
-#print(_path_list)
-#print("-" * 80)
-
-for _path in _path_list:
-    if _path is None: continue
-    _uc = _load_lib(_path, _lib.get(sys.platform, "libunicorn.so"))
-    if _uc is not None:
-        break
-
-# Try to search old unicorn1 library without SONAME
-if _uc is None:
-    for _path in _path_list:
-        if _path is None:
-            continue
-        
-        _uc = _load_lib(_path, "libunicorn.so")
-        if _uc is not None:
-            # In this case, show a warning for users
-            print("Found an old style dynamic library libunicorn.so, consider checking your installation", file=sys.stderr)
-            break
-
-if _uc is None:
-    raise ImportError("ERROR: fail to load the dynamic library.")
-
-__version__ = "%u.%u.%u" % (uc.UC_VERSION_MAJOR, uc.UC_VERSION_MINOR, uc.UC_VERSION_PATCH)
-
-# setup all the function prototype
-def _setup_prototype(lib, fname, restype, *argtypes):
-    try:
-        getattr(lib, fname).restype = restype
-        getattr(lib, fname).argtypes = argtypes
-    except AttributeError:
-        raise ImportError("ERROR: Fail to setup some function prototypes. Make sure you have cleaned your unicorn1 installation.")
-
-ucerr = ctypes.c_int
-uc_mode = ctypes.c_int
-uc_arch = ctypes.c_int
-uc_engine = ctypes.c_void_p
-uc_context = ctypes.c_void_p
-uc_hook_h = ctypes.c_size_t
+    import inspect
+    import os
+    import pkg_resources
+    import sys
+
+    loaded_dlls = set()
+
+    def __load_win_support(path: Path) -> None:
+        # Windows DLL in dependency order
+        all_dlls = (
+            'libwinpthread-1.dll',
+            'libgcc_s_seh-1.dll',
+            'libgcc_s_dw2-1.dll'
+        )
+
+        for dllname in all_dlls:
+            if dllname not in loaded_dlls:
+                lib_file = path / dllname
+
+                if str(path.parent) == '.' or lib_file.exists():
```

Comment:
`str(path.parent) == '.'` seems not equavelent to original logic? I suppose it should be `str(path)`.

Reply:
As far as I understood it, the original logic was trying to figure out whether this is a relative library path (i.e. either a filename without a parent directory, or just not an absolute path). Testing for parent == "." is the way to determine this. Did you encounter any problem loading the library this way? (I admit I haven't tested it thoroughtly)

I am not fully comfortable with how the library lookup code ended up, but considering the previous implementation I think it looks very nice after all..